snuba/migrations/validator.py

```diff
@@ -0,0 +1,199 @@
+import re
+from typing import Sequence, Union
+
+from snuba.clickhouse.native import ClickhousePool
+from snuba.clusters.cluster import ClickhouseClientSettings, get_cluster
+from snuba.migrations.migration import ClickhouseNodeMigration
+from snuba.migrations.operations import AddColumn, CreateTable, DropColumn, SqlOperation
+
+ENGINE_REGEX = r"^Distributed(\(.+\))$"
+
+
+class InvalidMigrationOrderError(Exception):
+    """
+    Raised when a migration ops are not in the correct order.
+    """
+
+    pass
+
+
+class NoDistClusterNodes(Exception):
+    """
+    The migration has no distributed cluster nodes, so we can't validate the order.
+    """
+
+    pass
+
+
+class DistributedEngineParseError(Exception):
+    """
+    Cant parser the distributed engine string value
+    """
+
+    pass
+
+
+def validate_migration_order(migration: ClickhouseNodeMigration) -> None:
+    """
+    Validates that the migration order is correct. Checks that the order of
+    AddColumn, CreateTable and DropColumn operations are correct with regards to being
+    applied on local and distributed tables.
+    """
+
+    def conflicts_ops(local_op: SqlOperation, dist_op: SqlOperation) -> bool:
+        if isinstance(local_op, CreateTable) and isinstance(dist_op, CreateTable):
+            return conflicts_create_table_op(local_op, dist_op)
+        elif isinstance(local_op, AddColumn) and isinstance(dist_op, AddColumn):
+            return conflicts_add_column_op(local_op, dist_op)
+        elif isinstance(local_op, DropColumn) and isinstance(dist_op, DropColumn):
+            return conflicts_drop_column_op(local_op, dist_op)
+        return False
+
+    def validate_add_col_or_create_table(
```

Comment:
is there a reason to keep these functions under the `validate_migration_order`? versus pulling them out to be at the top level? That would make it easier to test these functions in unit tests.

Reply:
Two reasons:
1. These functions were very specific to this method and I did want them exposed as part of the module API. It made sense to keep them private.
2. Test should test functionality rather than implementation. That way when we make changes to the implementation (which I anticipate we will do once we identify more conflicting ops), we can use the same tests